logfire/_internal/baggage.py

```diff
@@ -0,0 +1,101 @@
+from __future__ import annotations
+
+import warnings
+from collections.abc import Iterator
+from contextlib import contextmanager
+
+from opentelemetry import baggage, context
+from opentelemetry.sdk.trace import Span, SpanProcessor
+
+__all__ = (
+    'get_baggage',
+    'set_baggage',
+)
+
+from .utils import truncate_string
+
+get_baggage = baggage.get_all
+"""Get all OpenTelemetry baggage for the current context as a mapping of key/value pairs."""
+
+
+# Truncate strings to at most this length.
+# Thoughts behind this number:
+# - Nice and round.
+# - OTel baggage propagation limits total size to 8192 bytes, and each key/value pair is limited to 4096 bytes.
+# - UTF-8 can take up to 4 bytes per character.
+#    (the OTel code seems to actually measure characters, but it looks like it's supposed to be bytes)
+# - We don't want big attributes in every span.
+MAX_BAGGAGE_VALUE_LENGTH = 1000
+
+
+@contextmanager
+def set_baggage(**values: str) -> Iterator[None]:
+    """Context manager that attaches key/value pairs as OpenTelemetry baggage to the current context.
+
+    See the [Baggage documentation](https://logfire.pydantic.dev/docs/reference/advanced/baggage/) for more details.
+
+    Note: this function should always be used in a `with` statement; if you try to open and close it manually you may
+    run into surprises because OpenTelemetry Baggage is stored in the same contextvar as the current span.
+
+    Args:
+        values: The key/value pairs to attach to baggage. These should not be large or sensitive.
+            Strings longer than 1000 characters will be truncated with a warning.
+
+    Example usage:
+
+    ```python
+    from logfire import set_baggage
+
+    with set_baggage(my_id='123'):
+        # All spans opened inside this block will have baggage '{"my_id": "123"}'
+        with set_baggage(my_session='abc'):
+            # All spans opened inside this block will have baggage '{"my_id": "123", "my_session": "abc"}'
+            ...
+    ```
+    """
+    current_context = context.get_current()
+    for key, value in values.items():
+        if isinstance(value, str):  # type: ignore  # prevent error in line below if user ignores type hints
```

Comment:
What if it's not a string? I think we should emit a warning and do a best-effort to make it a string